torch_geometric/data/feature_store.py

```diff
@@ -0,0 +1,354 @@
+        sets the `index` of an :obj:`AttrView`:
+
+        .. code-block:: python
+
+            view = store.view(TensorAttr(group_name))
+            view.index = torch.Tensor([1, 2, 3])
+
+        """
+        if key in ['_attr', '_store']:
+            return super(AttrView, self).__setattr__(key, value)
+
+        TensorAttr.__setattr__(self._attr, key, value)
+
+    def __getitem__(
+        self,
+        index: IndexType,
+    ) -> Union['AttrView', FeatureTensorType]:
+        r"""Supports indexing the backing :obj:`TensorAttr` object by an
+        index or a slice. If the index operation results in a fully-specified
+        :obj:`AttrView`, a Tensor is returned. Otherwise, the :obj:`AttrView`
+        object is returned. The following operation returns a Tensor object
+        as a result of the index specification:
+
+        .. code-block:: python
+
+            store[group_name, attr_name][:]
+
+        """
+        self._attr.index = index
+        if self._attr.is_fully_specified():
+            return self._store.get_tensor(self._attr)
+        return self
+
+    def __call__(self) -> FeatureTensorType:
+        r"""Supports :obj:`AttrView` as a callable to force retrieval from
+        the currently specified attributes. In particular, this passes the
+        current :obj:`TensorAttr` object to a GET call, regardless of whether
+        all attributes have been specified. It returns the result of this
+        call. In particular, the following operation returns a Tensor by
+        performing a GET operation on the backing feature store:
+
+        .. code-block:: python
+
+            store[group_name, attr_name]()
+
+        """
+        return self._store.get_tensor(self._attr)
+
+    def __eq__(self, __o: object) -> bool:
+        r"""Compares two :obj:`AttrView` objects by checking equality of their
+        :obj:`FeatureStore` references and :obj:`TensorAttr` attributes."""
+        if not isinstance(__o, AttrView):
+            return False
+
+        return id(self._store) == id(__o._store) and self._attr == __o._attr
+
+    def __repr__(self) -> str:
+        return (f'{self.__class__.__name__}(store={self._store}, '
+                f'attr={self._attr})')
+
+
+class FeatureStore(MutableMapping):
+    def __init__(self, attr_cls: Any = TensorAttr):
+        r"""Initializes the feature store. Implementor classes can customize
+        the ordering and require nature of their :obj:`TensorAttr` tensor
+        attributes by subclassing :class:`TensorAttr` and passing the subclass
+        as `attr_cls`."""
+        super().__init__()
+        self._attr_cls = attr_cls
+
+    # Core (CRUD) #############################################################
+
+    @abstractmethod
+    def _put_tensor(self, tensor: FeatureTensorType, attr: TensorAttr) -> bool:
+        r"""Implemented by :obj:`FeatureStore` subclasses."""
+        pass
+
+    def put_tensor(self, tensor: FeatureTensorType, attr: TensorAttr) -> bool:
```

Comment:
```suggestion
    def put_tensor(self, tensor: FeatureTensorType, *args, **kwargs) -> bool:
```
This would be the syntax I would prefer. Which leads to
```python
store.put_tensor(tensor, group_name=..., attr_name=...)
```
We can do `self._attr_cls.cast(*args, **kwargs)` as well to cast this.